### PR TITLE
[verilator] Fix ELF processing

### PR DESCRIPTION
**Pull request by bcantrill**
As described in https://github.com/lowRISC/opentitan/issues/1245, the Ibex Verliator ELF processing does not exactly follow the algorithm for the "binary" target of objcopy (which it is intended to mimic). In particular, sections as denoted by program headers are laid down strictly sequentially, ignoring their specified `p_paddr` -- and resulting in a corrupted in-memory image for ELF images that have unordered program headers.

### CODE REVIEW

File: dv/verilator/memutil/cpp/verilator_memutil.cc [hunk 1]
```diff
@@ -228,9 +228,13 @@ bool VerilatorMemUtil::IsFileReadable(std::string filepath) const {
 bool VerilatorMemUtil::ElfFileToBinary(const std::string &filepath,
                                        uint8_t **data,
                                        size_t &len_bytes) const {
-  bool retval;
-  std::list<BufferDesc> buffers;
```
Comment (imphil) [inline dv/verilator/memutil/cpp/verilator_memutil.cc]: Can you remove the BufferDesc struct definition in the beginning of this file, it's now unused.

File: dv/verilator/memutil/cpp/verilator_memutil.cc [hunk 2]
```diff
@@ -228,9 +228,13 @@ bool VerilatorMemUtil::IsFileReadable(std::string filepath) const {
...
-  size_t offset = 0;
+  uint8_t *buf;
+  bool retval, any = false;
+  GElf_Phdr phdr;
+  GElf_Addr high = 0, low = (GElf_Addr)-1;
```
Comment (imphil) [inline dv/verilator/memutil/cpp/verilator_memutil.cc]: can you split this into two lines, and run clang-format on the whole change to fix the whitespace?

File: dv/verilator/memutil/cpp/verilator_memutil.cc [hunk 3]
```diff
@@ -271,48 +275,59 @@ bool VerilatorMemUtil::ElfFileToBinary(const std::string &filepath,
...
+
     if (phdr.p_type != PT_LOAD) {
-      std::cout << "Program header number " << i << "is not of type PT_LOAD."
-                << "Continue." << std::endl;
+      std::cout << "Program header number " << i << " is not of type PT_LOAD; "
+                << "ignoring." << std::endl;
       continue;
     }
+
+    if (phdr.p_filesz == 0)
```
Comment (imphil) [inline dv/verilator/memutil/cpp/verilator_memutil.cc]: We're trying to follow the [OpenTitan C/C++ style guide](https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) here which requires curly braces even for single-line statements. Can you add them here and in the statements below?

File: dv/verilator/memutil/cpp/verilator_memutil.cc [hunk 4]
```diff
@@ -271,48 +275,59 @@ bool VerilatorMemUtil::ElfFileToBinary(const std::string &filepath,
...
+    if (!any || phdr.p_paddr < low)
+      low = phdr.p_paddr;
+
+    if (!any || phdr.p_paddr + phdr.p_filesz > high)
+      high = phdr.p_paddr + phdr.p_filesz;
+
+    any = true;
+  }
+
+  buf = (uint8_t *)malloc(len_bytes = (high - low));
```
Comment (imphil) [inline dv/verilator/memutil/cpp/verilator_memutil.cc]: - can you please put the `len_bytes` assignment into a separate line?
- Can you add an `assert(buf)` after this statement?

File: dv/verilator/memutil/cpp/verilator_memutil.cc (hunk 3 above)
Comment (tomroberts-lowrisc) [reply dv/verilator/memutil/cpp/verilator_memutil.cc]: This isn't mentioned in the README at the moment. I'll add it and get clang-format added to CI.

File: dv/verilator/memutil/cpp/verilator_memutil.cc (hunk 1 above)
Comment (bcantrill) [reply dv/verilator/memutil/cpp/verilator_memutil.cc]: Good eyes! Fixed -- and thank you!

File: dv/verilator/memutil/cpp/verilator_memutil.cc (hunk 2 above)
Comment (bcantrill) [reply dv/verilator/memutil/cpp/verilator_memutil.cc]: Fixed

File: dv/verilator/memutil/cpp/verilator_memutil.cc (hunk 3 above)
Comment (bcantrill) [reply dv/verilator/memutil/cpp/verilator_memutil.cc]: Fixed -- and worth adding to [CONTRIBUTING.md](https://github.com/lowRISC/ibex/blob/master/CONTRIBUTING.md) as well?

File: dv/verilator/memutil/cpp/verilator_memutil.cc (hunk 4 above)
Comment (bcantrill) [reply dv/verilator/memutil/cpp/verilator_memutil.cc]: Fixed

**Comment by imphil** [comment]
Thanks! I'll update the copy of this repository in opentitan tomorrow to get this fixed there as well.

**Pull request merged**